velox/common/caching/SsdCache.cpp

```diff
@@ -0,0 +1,505 @@
+      ++numRegions_;
+      return true;
+    } else {
+      LOG(ERROR) << "Failed to grow cache file " << filename_ << " to "
+                 << newSize;
+    }
+  }
+  std::vector<int32_t> candidates;
+  int64_t scoreSum = 0;
+  for (int i = 0; i < numRegions_; ++i) {
+    if (regionPins_[i]) {
+      continue;
+    }
+    if (candidates.empty() || regionScore_[i] < scoreSum / candidates.size()) {
+      scoreSum += regionScore_[i];
+      candidates.push_back(i);
+    }
+  }
+  if (candidates.empty()) {
+    suspended_ = true;
+    return false;
+  }
+  std::sort(
+      candidates.begin(), candidates.end(), [&](int32_t left, int32_t right) {
+        return regionScore_[left] < regionScore_[right];
+      });
+  // Free up to 3 lowest score regions.
+  if (candidates.size() > 3) {
+    candidates.resize(3);
+  }
+  clearRegionEntriesLocked(candidates);
+  writableRegions_ = std::move(candidates);
+  suspended_ = false;
+  return true;
+}
+
+void SsdFile::clearRegionEntriesLocked(
+    const std::vector<int32_t>& regionIndices) {
+  // Remove all 'entries_' where the dependent points one of 'regionIndices'.
+  auto it = entries_.begin();
+  while (it != entries_.end()) {
+    auto region = regionIndex(it->second.offset());
+    if (std::find(regionIndices.begin(), regionIndices.end(), region) !=
+        regionIndices.end()) {
+      it = entries_.erase(it);
+    } else {
+      ++it;
+    }
+  }
+  for (auto region : regionIndices) {
+    // While the region is being filled it may get score from
+    // hits. When it is full, it will get a score boost to be a little
+    // ahead of the best.
+    regionScore_[region] = 0;
+    regionSize_[region] = 0;
+  }
+}
+
+void SsdFile::store(std::vector<CachePin>& pins) {
+  std::sort(pins.begin(), pins.end());
+  uint64_t total = 0;
+  for (auto& pin : pins) {
+    total += pin.entry()->size();
+  }
+  int32_t storeIndex = 0;
+  while (storeIndex < pins.size()) {
+    auto [offset, available] = getSpace(pins, storeIndex);
+    if (!available) {
+      // No space can be reclaimed. The pins are freed when the caller is freed.
+      return;
+    }
+    int32_t numWritten = 0;
+    int32_t bytes = 0;
+    std::vector<iovec> iovecs;
+    for (auto i = storeIndex; i < pins.size(); ++i) {
+      auto entrySize = pins[i].entry()->size();
+      if (bytes + entrySize > available) {
+        break;
+      }
+      pins[i].entry()->setSsdFile(this, offset);
+      addEntryToIovecs(*pins[i].entry(), iovecs);
+      bytes += entrySize;
+      ++numWritten;
+    }
+    VELOX_CHECK_GE(fileSize_, offset + bytes);
+    auto rc = folly::pwritev(fd_, iovecs.data(), iovecs.size(), offset);
+    if (rc != bytes) {
+      LOG(ERROR) << "Failed to write to SSD " << errno;
+      return;
+    }
+    {
+      std::lock_guard<std::mutex> l(mutex_);
+      for (auto i = storeIndex; i < storeIndex + numWritten; ++i) {
+        auto entry = pins[i].entry();
+        entry->setSsdFile(this, offset);
```

Comment:
Is this a duplicate of L313? pins[i].entry()->setSsdFile(this, offset);

Reading further, I see that 'offset' here is different from offset on L313. Should L313 be deleted?